src/modules/Game.class.js

```diff
@@ -5,64 +5,405 @@
+          }
+
+          if (moveTile > 0) {
+            if (!didJustMerged && merge) {
+              didJustMerged = true;
+
+              const newNumber = this.state[tile + moveTile][column] * 2;
+
+              this.state[tile + moveTile][column] = newNumber;
+
+              this.score += newNumber;
+            } else {
+              didJustMerged = false;
+
+              this.state[tile + moveTile][column] = this.state[tile][column];
+            }
+
+            this.state[tile][column] = 0;
+
+            didTilesMove = true;
+          }
+        }
+      }
+    }
+
+    if (didTilesMove) {
+      this.createRandomTile();
+    }
+
+    return didTilesMove;
+  }
+
+  saveState() {
+    return this.state.map((row) => row.slice());
+  }
+
+  createRandomTile() {
+    let row = this.randomNumber(3);
+
+    while (!this.state[row].includes(0)) {
+      row = this.randomNumber(3);
+    }
+
+    let cell = this.randomNumber(3);
+
+    while (this.state[row][cell] !== 0) {
+      cell = this.randomNumber(3);
+    }
+
+    const newState = this.state.map((currentRow) => currentRow.slice());
+
+    newState[row][cell] = this.generateCellValue();
+    this.state = newState;
+    this.printTiles();
+  }
+
+  randomNumber(max) {
+    return Math.round(Math.random() * max);
+  }
+
+  generateCellValue() {
+    const result = Math.random();
+
+    return result > 0.9 ? 4 : 2;
+  }
+
+  updateStatus() {
+    if (this.didPlayerWin()) {
+      this.status = Game.Status.win;
+    } else if (this.didPlayerLose()) {
+      this.status = Game.Status.lose;
+    }
+  }
+
+  didPlayerWin() {
+    for (const tile of this.state.flat()) {
+      if (tile === 2048) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  didPlayerLose() {
+    for (const tile of this.state.flat()) {
+      if (tile === 0) {
+        return false;
+      }
+    }
+
+    for (let row = 0; row <= 3; row++) {
+      for (let tile = 0; tile <= 2; tile++) {
+        if (this.state[row][tile] === this.state[row][tile + 1]) {
+          return false;
+        }
+      }
+    }
+
+    for (let column = 0; column <= 3; column++) {
+      for (let tile = 0; tile <= 2; tile++) {
+        if (this.state[tile][column] === this.state[tile + 1][column]) {
+          return false;
+        }
+      }
+    }
+
+    return true;
+  }
+
+  printTiles() {
+    const cells = document.getElementsByClassName('field-cell');
+    const flatState = this.state.flat();
+
+    for (let i = 0; i < flatState.length; i++) {
+      const currentCell = cells[i];
+      const currentState = flatState[i];
+
+      if (currentState > 0) {
+        if (currentCell !== undefined) {
+          currentCell.textContent = currentState;
+          currentCell.className = `field-cell field-cell--${currentState}`;
+        }
+      } else {
+        if (currentCell !== undefined) {
+          currentCell.textContent = '';
+          currentCell.className = 'field-cell';
+        }
+      }
+    }
+  }
```

Comment:
The `printTiles` method directly manipulates the DOM. This might cause issues if the method is called in a non-browser environment. Consider abstracting DOM interactions to improve testability and separation of concerns.